protocol/daemons/slinky/client/price_fetcher.go

```diff
@@ -0,0 +1,143 @@
+	priceFeedServiceClient api.PriceFeedServiceClient
+	grpcClient             daemontypes.GrpcClient
+	pricesSocket           string
+	pricesConn             *grpc.ClientConn
+	slinky                 oracleclient.OracleClient
+	logger                 log.Logger
+}
+
+// NewPriceFetcher creates a PriceFetcher.
+func NewPriceFetcher(
+	marketPairFetcher MarketPairFetcher,
+	grpcClient daemontypes.GrpcClient,
+	pricesSocket string,
+	slinky oracleclient.OracleClient,
+	logger log.Logger) PriceFetcher {
+	return &PriceFetcherImpl{
+		marketPairFetcher: marketPairFetcher,
+		grpcClient:        grpcClient,
+		pricesSocket:      pricesSocket,
+		slinky:            slinky,
+		logger:            logger,
+	}
+}
+
+// Start initializes the underlying connections of the PriceFetcher.
+func (p *PriceFetcherImpl) Start(ctx context.Context) error {
+	cancelCtx, cf := context.WithTimeout(ctx, SlinkyPriceServerConnectionTimeout)
+	defer cf()
+	pricesConn, err := p.grpcClient.NewGrpcConnection(cancelCtx, p.pricesSocket)
+	if err != nil {
+		return err
+	}
+	p.pricesConn = pricesConn
+	p.priceFeedServiceClient = api.NewPriceFeedServiceClient(p.pricesConn)
+	return p.slinky.Start(ctx)
+}
+
+// Stop closes all open connections.
+func (p *PriceFetcherImpl) Stop() {
+	if p.pricesConn != nil {
+		_ = p.grpcClient.CloseConnection(p.pricesConn)
+	}
+	_ = p.slinky.Stop()
+}
+
+// FetchPrices pulls prices from Slinky, translates the returned data format to dydx-compatible types,
+// and sends the price updates to the index price cache via the pricefeed server.
+// It uses the MarketPairFetcher to efficiently map between Slinky's CurrencyPair primary key and dydx's
+// MarketParam (or MarketPrice) ID.
+//
+// The markets in the index price cache will only have a single index price (from slinky).
+// This is because the sidecar pre-aggregates market data.
+func (p *PriceFetcherImpl) FetchPrices(ctx context.Context) error {
+	// get prices from slinky sidecar via GRPC
+	slinkyResponse, err := p.slinky.Prices(ctx, &types.QueryPricesRequest{})
+	if err != nil {
+		return err
+	}
+
+	// update the prices keeper w/ the most recent prices for the relevant markets
+	var updates []*api.MarketPriceUpdate
+	for currencyPairString, priceString := range slinkyResponse.Prices {
+		// convert currency-pair string (index) into currency-pair object
+		currencyPair, err := oracletypes.CurrencyPairFromString(currencyPairString)
+		if err != nil {
+			return err
+		}
+		p.logger.Info("turned price pair to currency pair",
+			"string", currencyPairString,
+			"currency pair", currencyPair.String())
+
+		// get the market id for the currency pair
+		id, err := p.marketPairFetcher.GetIDForPair(currencyPair)
+		if err != nil {
+			p.logger.Info("slinky client returned currency pair not found in MarketPairFetcher",
+				"currency pair", currencyPairString,
+				"error", err)
+			continue
+		}
+
+		// parse the price string into a uint64
+		price, err := strconv.ParseUint(priceString, 10, 64)
```

Comment:
nit: are we sure all slinky prices will fit in uint64s? the oracle-service deals w/ big.Ints which may not fit here

Reply:
I'm not positive it will fit...but the `MarketPriceUpdate` type contains `ExchangePrice` types which has a uint64 https://github.com/skip-mev/v4-chain/blob/49e060b07e3cd2c80460463682e8f1bda1d2ed86/proto/dydxprotocol/daemons/pricefeed/price_feed.proto#L27

So eventually it needs to fit in a uint64 type.